File: HTY360Player/GLProgram.m
//
//  GLProgram.m
//  HTY360Player
//
//  Created by  on 11/8/15.
//  Copyright © 2015 Hanton. All rights reserved.
//

#import "GLProgram.h"

@interface GLProgram()

- (BOOL)compileShader:(GLuint *)shader
                 type:(GLenum)type
               string:(NSString *)shaderString;

@end

@implementation GLProgram

#pragma mark - Init

- (id)initWithVertexShaderString:(NSString *)vShaderString
            fragmentShaderString:(NSString *)fShaderString {
    self = [super init];
    if (self) {
        self.attributes = [[NSMutableArray alloc] init];
        self.uniforms = [[NSMutableArray alloc] init];
        self.program = glCreateProgram();
        
        if (![self compileShader:&_vertShader
                            type:GL_VERTEX_SHADER
                          string:vShaderString])
            NSLog(@"Failed to compile vertex shader");
        
        if (![self compileShader:&_fragShader
                            type:GL_FRAGMENT_SHADER
                          string:fShaderString])
            NSLog(@"Failed to compile fragment shader");
        
        glAttachShader(_program, _vertShader);
        glAttachShader(_program, _fragShader);
    }
    
    return self;
}

- (id)initWithVertexShaderFilename:(NSString *)vShaderFilename
            fragmentShaderFilename:(NSString *)fShaderFilename {
    NSString *vertShaderPathname = [[NSBundle mainBundle] pathForResource:vShaderFilename ofType:@"vsh"];
    NSString *vertexShaderString = [NSString stringWithContentsOfFile:vertShaderPathname
                                                             encoding:NSUTF8StringEncoding
                                                                error:nil];
    
    NSString *fragShaderPathname = [[NSBundle mainBundle] pathForResource:fShaderFilename ofType:@"fsh"];
    NSString *fragmentShaderString = [NSString stringWithContentsOfFile:fragShaderPathname
                                                               encoding:NSUTF8StringEncoding
                                                                  error:nil];
    
    if ((self = [self initWithVertexShaderString:vertexShaderString fragmentShaderString:fragmentShaderString])) {
        
    }
    
    return self;
}

#pragma mark - Dealloc

- (void)dealloc {
    if (self.vertShader) {
        glDeleteShader(self.vertShader);
    }
    
    if (self.fragShader) {
        glDeleteShader(self.fragShader);
    }
    
    if (self.program) {
        glDeleteProgram(self.program);
        self.program = 0;
    }
}

#pragma mark - Private Method

- (BOOL)compileShader:(GLuint *)shader
                 type:(GLenum)type
               string:(NSString *)shaderString {
    GLint status;
    const GLchar *source;
    
    source = (GLchar *)[shaderString UTF8String];
    if (!source) {
        NSLog(@"Failed to load shader: %@", shaderString);
        return NO;
    }
    
    *shader = glCreateShader(type);
    glShaderSource(*shader, 1, &source, NULL);
    glCompileShader(*shader);
    
    glGetShaderiv(*shader, GL_COMPILE_STATUS, &status);
    
    if (status != GL_TRUE) {
        GLint logLength;
        glGetShaderiv(*shader, GL_INFO_LOG_LENGTH, &logLength);
        if (logLength > 0) {
            GLchar *log = (GLchar *)malloc(logLength);
            glGetShaderInfoLog(*shader, logLength, &logLength, log);
            NSLog(@"Shader compile log:\n%s", log);
            free(log);
        }
    }
    
    return status == GL_TRUE;
}

#pragma mark - Public Method

- (void)addAttribute:(NSString *)attributeName {
    if (![self.attributes containsObject:attributeName]) {
        [self.attributes addObject:attributeName];
        glBindAttribLocation(self.program,
                             (GLuint)[self.attributes indexOfObject:attributeName],
                             [attributeName UTF8String]);
    }
}

- (GLuint)attributeIndex:(NSString *)attributeName {
    return (GLuint)[self.attributes indexOfObject:attributeName];
}

- (GLuint)uniformIndex:(NSString *)uniformName {
    return glGetUniformLocation(self.program, [uniformName UTF8String]);
}

- (BOOL)link {
    GLint status;
    glLinkProgram(self.program);
    
    glGetProgramiv(self.program, GL_LINK_STATUS, &status);
    if (status == GL_FALSE)
        return NO;
    
    if (self.vertShader) {
        glDeleteShader(self.vertShader);
        self.vertShader = 0;
    }
    if (self.fragShader) {
        glDeleteShader(self.fragShader);
        self.fragShader = 0;
    }
    
    return YES;
}

- (void)use {
    glUseProgram(self.program);
}

@end

File: DemoApp/DemoApp/Shader/GLProgram.swift
//
//  ViewController.swift
//  Fisheye
//
//  Created by Hanton Yang on 2/6/23.
//

import Foundation
import OpenGLES.ES3

class GLProgram {
    private var programHandle: GLuint = 0
    private var vertexShader: GLuint = 0
    private var fragmentShader: GLuint = 0

    func compileShaders(vertexShaderName: String, fragmentShaderName: String) -> GLuint {
        programHandle = glCreateProgram()

        if !compileShader(&vertexShader, type: GLenum(GL_VERTEX_SHADER), file: Bundle.main.path(forResource: vertexShaderName, ofType: "glsl")!) {
            print("vertex shader failure")
        }

        if !compileShader(&fragmentShader, type: GLenum(GL_FRAGMENT_SHADER), file: Bundle.main.path(forResource: fragmentShaderName, ofType: "glsl")!) {
            print("fragment shader failure")
        }

        glAttachShader(programHandle, vertexShader)
        glAttachShader(programHandle, fragmentShader)

        if !link() {
            print("link failure")
        }

        return programHandle
    }

    private func link() -> Bool {
        var status: GLint = 0

        glLinkProgram(programHandle)
        glGetProgramiv(programHandle, GLenum(GL_LINK_STATUS), &status)

        if status == GL_FALSE {
            return false
        }

        if vertexShader > 0 {
            glDeleteShader(vertexShader)
            vertexShader = 0
        }

        if fragmentShader > 0 {
            glDeleteShader(fragmentShader)
            fragmentShader = 0
        }

        return true
    }

    private func compileShader(_ shader: inout GLuint, type: GLenum, file: String) -> Bool {
        var status: GLint = 0
        var source: UnsafePointer<Int8>

        do {
            source = try NSString(contentsOfFile: file, encoding: String.Encoding.utf8.rawValue).utf8String!
        } catch {
            print("failed to load shader")
            return false
        }

        var castSource: UnsafePointer<GLchar>? = UnsafePointer<GLchar>(source)

        shader = glCreateShader(type)
        glShaderSource(shader, 1, &castSource, nil)
        glCompileShader(shader)
        glGetShaderiv(shader, GLenum(GL_COMPILE_STATUS), &status)

        if status == GL_FALSE {
            glDeleteShader(shader)
            return false
        }

        return true
    }
}
